### PR TITLE
feat: support get plugin schema based on schema_type

### PR DESCRIPTION
**Pull request by liuxiran**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] Bugfix
- [x] New feature provided
- [ ] Improve performance

- Related issues
https://github.com/apache/apisix-dashboard/issues/605

here I have two requests for help

1. after checked the lib `github.com/shiningrush/droplet`, it seems that **there is  no api to get query from request url**, so there is a `TODO mark` in 
 https://github.com/liuxiran/apisix-dashboard/blob/47348773e1e7c90250a0145053a1c3ee29c30e54/api/internal/handler/plugin/plugin.go#L51
needs @ShiningRush `s help when you are free ,many thx~!

1. An error occurred while running `api/build-tools/schema-sync.sh master` to update schema.json, here is the error msg
```shell
lua: ./apisix/plugins/api-breaker.lua:25: attempt to index a nil value (field 'shared')
stack traceback:
	./apisix/plugins/api-breaker.lua:25: in main chunk
	[C]: in function 'require'
	schema-sync.lua:128: in main chunk
	[C]: in ?
sync success:
/home/liuxiran/product/apisix-dashboard/api/conf/schema.json
```
and the `api-break.lua:25` is 
https://github.com/apache/apisix/blob/f2f5ab97663805509baf3355c37336ffa557b68c/apisix/plugins/api-breaker.lua#L25
it seems  to be due to a lack of ngx dependency, so how to add this dependency?
needs @nic-chen `s help, when you are free, many thx~!

### CODE REVIEW

**Comment by juzhiyuan** [comment]
The Sync schema issue needs @nic-chen’s help :) Please be patient 💯

**Comment by liuxiran** [comment]
> The Sync schema issue needs @nic-chen’s help :) Please be patient

ok, and I will deal with the ci error after the Q1 be solved :)

**Comment by nic-chen** [comment]
@liuxiran 

1. please refer to:
https://github.com/apache/apisix-dashboard/blob/master/api/internal/handler/consumer/consumer.go#L76-L87

2.
add  
```lua
ngx.shared = {
    "plugin-api-breaker" = {}
}
```
to line 81 in `schema-sync.lua` and have a try.

**Comment by ShiningRush** [comment]
If you tag the input field with`auto_read:"key,source"`, it will read it from query string automatically.For example:
```
type GetInput struct {
	Name string `auto_read:"name,query" validate:"required"`
}
```
Framework will read query string `name` to `GetInput.Name`,
`source` could be:
- path: if your path is like `/resources/{id}/child-resources`, it will read the real string of `{id}`
- query: read from query string

Emmm, I will append document for `droplet`.

**Comment by liuxiran** [comment]
- [ ] api/internal/core/store/validate.go Validate函数需要增加 `schemaType` 参数

- [ ] 对应调用到的地方要做传参

**Comment by liuxiran** [comment]
> @liuxiran
> 
> 1. please refer to:
>    https://github.com/apache/apisix-dashboard/blob/master/api/internal/handler/consumer/consumer.go#L76-L87
> 
> add
> 
> ```lua
> ngx.shared = {
>     ["plugin-api-breaker"] = {}
> }
> ```
> 
> to line 81 in `schema-sync.lua` and have a try.

thx for your  help, already updated pr :)

**Review by juzhiyuan — APPROVED** [review]
LGTM from comments

**Comment by liuxiran** [comment]
> If you tag the input field with`auto_read:"key,source"`, it will read it from query string automatically.For example:
> 
> ```
> type GetInput struct {
> 	Name string `auto_read:"name,query" validate:"required"`
> }
> ```
> 
> Framework will read query string `name` to `GetInput.Name`,
> `source` could be:
> 
> * path: if your path is like `/resources/{id}/child-resources`, it will read the real string of `{id}`
> * query: read from query string
> 
> Emmm, I will append document for `droplet`.

thx for your instructions, 

> If you tag the input field with`auto_read:"key,source"`, it will read it from query string automatically.For example:
> 
> ```
> type GetInput struct {
> 	Name string `auto_read:"name,query" validate:"required"`
> }
> ```
> 
> Framework will read query string `name` to `GetInput.Name`,
> `source` could be:
> 
> * path: if your path is like `/resources/{id}/child-resources`, it will read the real string of `{id}`
> * query: read from query string
> 
> Emmm, I will append document for `droplet`.

thx for your instructions, as a important lib in 2.0 manager-api, a doc would be very helpful for a beginner 😍 @ShiningRush

**Comment by moonming** [comment]
@membphis please take a look

File: api/build-tools/schema-sync.lua [hunk 1]
```diff
@@ -78,6 +78,9 @@ ngx.timer = {}
 ngx.location = {}
 ngx.socket = {}
 ngx.re.gmatch = empty_function
+ngx.shared = {
+    ["plugin-api-breaker"] = {}
```
Comment (membphis) [inline api/build-tools/schema-sync.lua:82]: we can use `__index` to mock all of the share dicts.

Comment (membphis) [reply api/build-tools/schema-sync.lua:82]: I have submitted an issue: https://github.com/apache/apisix-dashboard/issues/671, we can fix this later.

File: api/conf/schema.json [hunk 2]
```diff
@@ -1 +1 @@
...
...host for upstream","pattern":"^[0-9a-zA-Z-.]+$","type":"string"},"regex_uri":{"description":"new uri that substitute from client uri for upstream, lower priority than uri property","type":"array","minItems":2,"items":{"description":"regex uri","type":"string"},"maxItems":2},"headers":{"description":"new headers for request","minProperties":1,"type":"object"},"uri":{"description":"new uri for upstream","maxLength":4096,"type":"string","pattern":"^\\/.*","minLength":1}}}},"batch-requests":{"schema":{"type":"object","additionalProperties":false}},"prometheus":{"schema":{"type":"object","additionalProperties":false}},"echo":{"schema":{"anyOf":[{"required":["before_body"]},{"required":["body"]},{"required":["after_body"]}],"properties":{"after_body":{"description":"body after the modification of filter phase.","type":"string"},"headers":{"description":"new headers for response","minProperties":1,"type":"object"},"before_body":{"description":"body before the filter phase.","type":"string"},"auth_value":{"description":"auth value","type":"string"},"body":{"description":"body to replace upstream response.","type":"string"}},"type":"object","minProperties":1,"additionalProperties":false}}}}
```
Comment (membphis) [inline api/conf/schema.json:1]: it is hard to confirm the new result is the correct one.

I think we can use YAML or formatted JSON style. JSON or YAML, the results need to be in order, which makes it easier to compare changes.

Anyway, I'll create a new issue for this, we can implement this feature in a new PR.

```json
{
	"traceSegmentId": "4560dd63-1030-4905-88e1-8496223f1925",
	"traceId": "4560dd63-1030-4905-88e1-8496223f1925",
	"service": "APISIX",
	"spans": [{
		"peer": "upstream service",
		"spanId": 1,
		"spanLayer": "Http",
		"spanType": "Exit",
		"startTime": 1603598158143,
		"endTime": 1603598158143,
		"operationName": "/hello",
		"componentId": 6000,
		"isError": false,
		"tags": [{
			"key": "http.status",
			"value": 200
		}],
		"parentSpanId": 0
	}, {
		"spanId": 0,
		"spanLayer": "Http",
		"spanType": "Entry",
		"startTime": 1603598158143,
		"endTime": 1603598158143,
		"operationName": "/hello",
		"componentId": 6000,
		"isError": false,
		"tags": [{
			"key": "http.method",
			"value": "GET"
		}, {
			"key": "http.params",
			"value": "http://127.0.0.1/hello"
		}, {
			"key": "http.status",
			"value": "200"
		}],
		"parentSpanId": -1
	}],
	"serviceInstance": "APISIX Instance Name"
}
```

```yaml
traceSegmentId: 4560dd63-1030-4905-88e1-8496223f1925
traceId: 4560dd63-1030-4905-88e1-8496223f1925
service: APISIX
spans:
  - peer: upstream service
    spanId: 1
    spanLayer: Http
    spanType: Exit
    startTime: 1603598158143
    endTime: 1603598158143
    operationName: /hello
    componentId: 6000
    isError: false
    tags:
      - key: http.status
        value: 200
    parentSpanId: 0
  - spanId: 0
    spanLayer: Http
    spanType: Entry
    startTime: 1603598158143
    endTime: 1603598158143
    operationName: /hello
    componentId: 6000
    isError: false
    tags:
      - key: http.method
        value: GET
      - key: http.params
        value: 'http://127.0.0.1/hello'
      - key: http.status
        value: '200'
    parentSpanId: -1
serviceInstance: APISIX Instance Name
```

File: api/internal/core/store/validate.go [hunk 3]
```diff
@@ -230,14 +231,21 @@ func (v *APISIXJsonSchemaValidator) Validate(obj interface{}) error {
...
 	//check plugin json schema
-	plugins := getPlugins(obj)
+	plugins, schemaType := getPlugins(obj)
+	//fix lua json.encode transform lua{properties={}} to json{"properties":[]}
+	reg := regexp.MustCompile(`\"properties\":\[\]`)
 	if plugins != nil {
 		for pluginName, pluginConf := range plugins {
-			schemaDef := conf.Schema.Get("plugins." + pluginName).String()
+			var schemaDef string
+			schemaDef = conf.Schema.Get("plugins." + pluginName + "." +schemaType).String()
```
Comment (membphis) [inline api/internal/core/store/validate.go:240]: style: `"." +schemaType` -> `"." + schemaType`

File: api/internal/core/store/validate.go [hunk 4]
```diff
@@ -230,14 +231,21 @@ func (v *APISIXJsonSchemaValidator) Validate(obj interface{}) error {
 	}
 
 	//check plugin json schema
-	plugins := getPlugins(obj)
+	plugins, schemaType := getPlugins(obj)
+	//fix lua json.encode transform lua{properties={}} to json{"properties":[]}
```
Comment (membphis) [inline api/internal/core/store/validate.go:235]: we can avoid this hack, Lua `json.encode` can do this in a good way.

@nic-chen here is the API, take a look: https://github.com/openresty/lua-cjson/#encode_empty_table_as_object

you can help @liuxiran to fix this

File: api/internal/handler/plugin/plugin_test.go [hunk 5]
```diff
@@ -55,4 +55,48 @@ func TestPlugin(t *testing.T) {
...
+	/*
+	 get plugin schema with schema_type: consumer
+	 plugin has consumer_schema
+	 return plugin`s consumer_schema
+	*/
+	input3 := &GetInput{}
+	reqBody = `{
+		"name": "jwt-auth",
+		"schema_type": "consumer"
+  }`
```
Comment (membphis) [inline api/internal/handler/plugin/plugin_test.go:68]: bad indentation, please fix similar points

File: api/internal/handler/plugin/plugin_test.go [hunk 6]
```diff
@@ -55,4 +55,48 @@ func TestPlugin(t *testing.T) {
 	ctx.SetInput(input2)
 	val, _ = handler.Schema(ctx)
 	assert.Nil(t, val)
+
+	/*
+	 get plugin schema with schema_type: consumer
+	 plugin has consumer_schema
+	 return plugin`s consumer_schema
+	*/
+	input3 := &GetInput{}
```
Comment (membphis) [inline api/internal/handler/plugin/plugin_test.go:64]: can we reuse one variable, eg: `input` ?

**Review by membphis — COMMENTED** [review]
for the test case, I would like this style: https://github.com/apache/apisix-dashboard/blob/v2.0/api/test/e2e/route_test.go#L117

it seems clearer.

File: api/conf/schema.json (hunk 2 above)
Comment (membphis) [reply api/conf/schema.json:1]: new issue: https://github.com/apache/apisix-dashboard/issues/672

let us discuss in the issue.

**Review by moonming — COMMENTED** [review]
What problem is this PR trying to solve? Sorry, I am not familiar with this part of the logic.
I have some questions:
1. Why dashboard repo has Lua code?
2. If Apache APISIX adds or modifies the plug-in schema, do we need to modify the code here?

**Comment by juzhiyuan** [comment]
> What problem is this PR trying to solve? Sorry, I am not familiar with this part of the logic.
> I have some questions:
> 
> 1. Why dashboard repo has Lua code?
> 2. If Apache APISIX adds or modifies the plug-in schema, do we need to modify the code here?

Related issue https://github.com/apache/apisix-dashboard/issues/605

This PR aims to fetch the plugin's schema type.

> Why dashboard repo has Lua code?

@nic-chen uses those codes to sync schema from APISIX.

> If Apache APISIX adds or modifies the plug-in schema, do we need to modify the code here?

Yes, and this way should be removed and use other good solutions.

File: api/conf/schema.json (hunk 2 above)
Comment (liuxiran) [reply api/conf/schema.json:1]: got it~!

File: api/build-tools/schema-sync.lua (hunk 1 above)
Comment (liuxiran) [reply api/build-tools/schema-sync.lua:82]: got it~!

File: api/internal/core/store/validate.go (hunk 3 above)
Comment (liuxiran) [reply api/internal/core/store/validate.go]: done~!

File: api/internal/handler/plugin/plugin_test.go (hunk 5 above)
Comment (liuxiran) [reply api/internal/handler/plugin/plugin_test.go]: done~!

File: api/internal/handler/plugin/plugin_test.go (hunk 6 above)
Comment (liuxiran) [reply api/internal/handler/plugin/plugin_test.go]: done~!

**Review by membphis — APPROVED** [review]
LGTM

**Comment by membphis** [comment]
@liuxiran You need to resolve the conflicting files first ^_^

**Pull request merged**